pkg/template/controller/templateinstance_controller.go

```diff
@@ -0,0 +1,170 @@
+	templateapiv1 "github.com/openshift/origin/pkg/template/api/v1"
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/meta"
+	"k8s.io/kubernetes/pkg/api/meta/metatypes"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/auth/user"
+	"k8s.io/kubernetes/pkg/client/cache"
+	kclientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/pkg/client/restclient"
+	"k8s.io/kubernetes/pkg/kubectl/resource"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+type TemplateInstanceController struct {
+	restconfig restclient.Config
+	oc         *client.Client
+	controller *cache.Controller
+}
+
+func NewTemplateInstanceController(restconfig restclient.Config, oc *client.Client) *TemplateInstanceController {
+	c := TemplateInstanceController{restconfig: restconfig, oc: oc}
+
+	_, c.controller = cache.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(options kapi.ListOptions) (runtime.Object, error) {
+				return c.oc.TemplateInstances(kapi.NamespaceAll).List(options)
+			},
+			WatchFunc: func(options kapi.ListOptions) (watch.Interface, error) {
+				return c.oc.TemplateInstances(kapi.NamespaceAll).Watch(options)
+			},
+		},
+		&templateapi.TemplateInstance{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				c.stateMachine(obj.(*templateapi.TemplateInstance))
+			},
+			UpdateFunc: func(_, obj interface{}) {
+				c.stateMachine(obj.(*templateapi.TemplateInstance))
+			},
+			DeleteFunc: func(obj interface{}) {
+			},
+		},
+	)
+
+	return &c
+}
+
+func (c *TemplateInstanceController) Run(stop <-chan struct{}) {
+	c.controller.Run(stop)
+}
+
+func (c *TemplateInstanceController) stateMachine(ti *templateapi.TemplateInstance) error {
+	for _, condition := range ti.Status.Conditions {
+		if condition.Type == templateapi.TemplateInstanceCreated && condition.Status == kapi.ConditionTrue ||
+			condition.Type == templateapi.TemplateInstanceFailed && condition.Status == kapi.ConditionTrue {
+			return nil
+		}
+	}
+
+	err := c.provision(ti)
+	if err == nil {
+		ti.Status.Conditions = []templateapi.TemplateInstanceCondition{
+			{
+				Type:               templateapi.TemplateInstanceCreated,
+				Status:             kapi.ConditionTrue,
+				LastTransitionTime: unversioned.Now(),
+				Reason:             "Created",
+			},
+		}
+
+	} else {
+		ti.Status.Conditions = []templateapi.TemplateInstanceCondition{
+			{
+				Type:               templateapi.TemplateInstanceFailed,
+				Status:             kapi.ConditionTrue,
+				LastTransitionTime: unversioned.Now(),
+				Reason:             "Failed",
+				Message:            err.Error(),
+			},
+		}
+	}
+
+	_, err = c.oc.TemplateInstances(ti.Namespace).Update(ti)
+	return err
+}
+
+func (c *TemplateInstanceController) provision(ti *templateapi.TemplateInstance) error {
+	config := c.restconfig
+	config.WrapTransport = func(rt http.RoundTripper) http.RoundTripper {
```

Comment:
@deads2k this is where the impersonation is actually carried out - the intention is that the controller impersonates the username from the configuration struct to create all the template items